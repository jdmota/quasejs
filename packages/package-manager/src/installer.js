// @flow

import { read as readPkg } from "./pkg";
import { read as readLockfile, write as writeLockfile } from "./lockfile";
import resolve from "./resolve";
import { Tree } from "./resolution";
import Store from "./store";

const path = require( "path" );

export type InstallOptions = {
  store: string,
  cache: string,
  offline: ?boolean,
  preferOffline: ?boolean,
  preferOnline: ?boolean,
  flat: ?boolean,
  update: ?boolean,
  check: ?boolean
};

type DepType = "deps" | "devDeps" | "optionalDeps";

type Deps = { [name: string]: { savedVersion: string, resolved: string, i: number } };

// TODO refactor out lockfile logic, and check for validity of it

// [ name, version, resolution, integrity, deps ]
export type Entry = [string, string, string, string, number[]];

const LOCK_VERSION = "1"; // TODO deal with different versions

type Lockfile = {
  v: string,
  resolutions: Entry[],
  deps: Deps,
  devDeps: Deps,
  optionalDeps: Deps,
};

/* eslint no-await-in-loop: 0 */

export default async function( folder: string, _opts: Object ) {

  const opts: InstallOptions = Object.assign( {}, _opts );
  if ( opts.store == null ) {
    opts.store = Store.DEFAULT;
  }
  if ( opts.offline == null && opts.preferOnline == null ) {
    opts.preferOffline = true;
  }
  opts.cache = path.join( opts.store, "cache" );

  const configPromises = [ readPkg( folder ), readLockfile( folder ) ];
  const pkg = await configPromises[ 0 ];
  const lockfile = ( await configPromises[ 1 ]: Lockfile );

  const newLockfile: Lockfile = {
    v: LOCK_VERSION,
    resolutions: [],
    deps: {},
    devDeps: {},
    optionalDeps: {}
  };

  const store = new Store();
  const tree = new Tree();

  const promises = [];

  async function installFromLock( index ) {
    const [ name, version, resolved, integrity ] = lockfile.resolutions[ index ];
    const obj = { name, version, resolved, integrity };

    const res = await tree.createResolution( obj, async set => {
      const extraction = store.extract( obj.resolved, opts );
      const promises = [];

      const indexes = lockfile.resolutions[ index ][ 4 ];
      for ( const i of indexes ) {
        promises.push( installFromLock( i ) );
      }

      for ( const p of promises ) {
        set.add( await p );
      }
      await extraction;
    } );

    await store.createResolution( res, opts );
    return res;
  }

  async function install( obj ) {
    const res = await tree.createResolution( obj, async set => {
      const extraction = store.extract( obj.resolved, opts );
      const promises = [];

      for ( const name in obj.deps ) {
        promises.push(
          resolve( name, obj.deps[ name ], opts ).then( install )
        );
      }

      for ( const p of promises ) {
        set.add( await p );
      }
      await extraction;
    } );

    await store.createResolution( res, opts );
    return res;
  }

  const allDeps = [];

  async function resolveDep( name: string, version: string, depType: DepType ) {
    if ( !opts.update ) {
      const deps = lockfile[ depType ] || {};
      const { savedVersion, resolved, i } = deps[ name ] || {};
      if ( savedVersion === version ) {
        newLockfile[ depType ][ name ] = deps[ name ];
        allDeps.push( resolved );
        return installFromLock( i );
      }
    }

    const obj = await resolve( name, version, opts );
    newLockfile[ depType ][ name ] = { savedVersion: version, resolved: obj.resolved, i: -1 };
    allDeps.push( obj.resolved );
    return install( obj );
  }

  for ( const name in pkg.dependencies ) {
    promises.push( resolveDep( name, pkg.dependencies[ name ], "deps" ) );
  }

  for ( const name in pkg.devDependencies ) {
    promises.push( resolveDep( name, pkg.devDependencies[ name ], "devDeps" ) );
  }

  await Promise.all( promises );
  await store.linkNodeModules( folder, await tree.extractDeps( allDeps ), opts, true );

  const map: Map<string, number> = new Map();
  tree.generate( newLockfile.resolutions, map );

  for ( const name in newLockfile.deps ) {
    // $FlowFixMe
    newLockfile.deps[ name ].i = map.get( newLockfile.deps[ name ].resolved );
  }

  for ( const name in newLockfile.devDeps ) {
    // $FlowFixMe
    newLockfile.devDeps[ name ].i = map.get( newLockfile.devDeps[ name ].resolved );
  }

  await writeLockfile( folder, newLockfile );

}
