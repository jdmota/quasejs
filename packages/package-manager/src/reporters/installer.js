// @flow
import { type Warning } from "../types";
import { BaseReporter } from "./base";

/* eslint-disable no-console */

const logSymbols = require( "log-symbols" );

export class InstallReporter extends BaseReporter {

  jobsTotal: number;
  jobsDone: number;
  phase: string;

  constructor() {
    super( "Installing..." );
    this.jobsTotal = 0;
    this.jobsDone = 0;
    this.phase = "";
  }

  resolutionStart( count: ?number ) {
    this.jobStart( "Resolution", count );
  }

  resolutionMore( count: ?number ) {
    this.jobMore( count );
  }

  resolutionUpdate() {
    this.jobUpdate();
  }

  extractionStart( count: ?number ) {
    this.jobStart( "Extraction", count );
  }

  extractionMore( count: ?number ) {
    this.jobMore( count );
  }

  extractionUpdate() {
    this.jobUpdate();
  }

  linkingStart( count: ?number ) {
    this.jobStart( "Linking", count );
  }

  linkingMore( count: ?number ) {
    this.jobMore( count );
  }

  linkingUpdate() {
    this.jobUpdate();
  }

  localLinkingStart( count: ?number ) {
    this.jobStart( "Local linking", count );
  }

  localLinkingMore( count: ?number ) {
    this.jobMore( count );
  }

  localLinkingUpdate() {
    this.jobUpdate();
  }

  folder( { folder }: { folder: string } ) {
    console.log( `${logSymbols.info} Project folder: ${folder}` );
  }

  lockfile( { reusing }: { reusing: boolean } ) {
    if ( reusing ) {
      console.log( `${logSymbols.info} Will reuse lockfile.` );
    } else {
      console.log( `${logSymbols.info} Will generate new lockfile.` );
    }
  }

  jobStart( phase: string, count: ?number ) {
    this.jobsTotal = count == null ? 0 : count;
    this.jobsDone = 0;
    this.phase = phase;
    this.spinner.text = `${phase} - Progress: 0/${this.jobsTotal}`;
  }

  jobMore( count: ?number ) {
    this.jobsTotal += count == null ? 1 : count;
    this.spinner.text = `${this.phase} - Progress: ${this.jobsDone}/${this.jobsTotal}`;
  }

  jobUpdate() {
    this.jobsDone++;
    this.spinner.text = `${this.phase} - Progress: ${this.jobsDone}/${this.jobsTotal}`;
  }

  updateLockfile() {
    this.spinner.text = `Updating lockfile...`;
  }

}

export class InstallReporterNoop extends InstallReporter {
  start() {}
  log( _: string ) {}
  error( _: Object ) {}
  warning( _: Warning ) {}
  done() {}

  resolutionStart( _: ?number ) {}
  resolutionMore( _: ?number ) {}
  resolutionUpdate() {}
  extractionStart( _: ?number ) {}
  extractionMore( _: ?number ) { }
  extractionUpdate() {}
  linkingStart( _: ?number ) {}
  linkingMore( _: ?number ) {}
  linkingUpdate() {}
  localLinkingStart( _: ?number ) {}
  localLinkingMore( _: ?number ) {}
  localLinkingUpdate() {}
  folder( _: { folder: string } ) {}
  lockfile( _: { reusing: boolean } ) {}
  jobStart( _: string, __: ?number ) {}
  jobMore( _: ?number ) {}
  jobUpdate() {}
  updateLockfile() {}
}
