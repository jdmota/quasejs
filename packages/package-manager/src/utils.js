// @flow

const crypto = require( "crypto" );
const fs = require( "fs-extra" );
const klaw = require( "klaw" );

export function isObject( obj: mixed ): boolean {
  return obj != null && typeof obj === "object";
}

export function isEmpty( obj: Object ) {
  for ( const key in obj ) {
    return false;
  }
  return true;
}

export function mapGet<K, V>( map: Map<K, V>, key: K ): V {
  // $FlowIgnore
  return map.get( key );
}

export function hash( input: string ) {
  return crypto.createHash( "md5" ).update( input ).digest( "hex" ).slice( 0, 10 );
}

export async function read( p: string ): Promise<string> {
  try {
    return await fs.readFile( p, "utf8" );
  } catch ( e ) {
    if ( e.code === "ENOENT" ) {
      return "";
    }
    throw e;
  }
}

export async function crawl( folder: string, mapper: Function ): Promise<mixed> {
  const promises = [];
  await new Promise( ( resolve, reject ) => {
    klaw( folder )
      .on( "data", item => {
        const p = mapper( item );
        if ( p ) {
          promises.push( p );
        }
      } )
      .on( "error", reject )
      .on( "end", resolve );
  } );
  return Promise.all( promises );
}
