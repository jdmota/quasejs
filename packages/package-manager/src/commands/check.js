// @flow
import { CheckReporter } from "../reporters/check";
import { type Name, toStr } from "../types";
import { error, readJSON } from "../utils";
import { read as readPkg } from "../pkg";
import { shouldReuse as shouldReuseLockfile, read as readLockfile } from "../lockfile";

const path = require( "path" );

function compare( a, b, type ) {

  for ( const nameStr in a ) {
    // $FlowIgnore
    const name: Name = nameStr;

    if ( !b[ name ] ) {
      throw error( `Package.json has ${nameStr} but lockfile has not (on ${type}).` );
    }
    if ( a[ name ] !== b[ name ].savedVersion ) {
      throw error(
        `Package.json has ${nameStr} with version ${a[ name ]} but lockfile has version ${toStr( b[ name ].savedVersion )} (on ${type}).`
      );
    }
  }

  for ( const nameStr in b ) {
    // $FlowIgnore
    const name: Name = nameStr;

    if ( !a[ name ] ) {
      throw error( `Lockfile has ${nameStr} but package.json has not (on ${type}).` );
    }
  }

}

async function integrity( folder, lockfile ) {

  const checks = [];

  for ( const nameStr in lockfile.deps ) {
    // $FlowIgnore
    const name: Name = nameStr;

    const resolvedObj = lockfile.deps[ name ];
    const idx = resolvedObj.i;
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${nameStr}@${toStr( version )}`, name, hash } );
  }

  for ( const nameStr in lockfile.devDeps ) {
    // $FlowIgnore
    const name: Name = nameStr;

    const resolvedObj = lockfile.devDeps[ name ];
    const idx = resolvedObj.i;
    const version = lockfile.resolutions[ idx ][ 1 ];
    const hash = lockfile.resolutions[ idx ][ 3 ];
    checks.push( { pkg: `${nameStr}@${toStr( version )}`, name, hash } );
  }

  return Promise.all(
    checks.map( async( { pkg, name, hash } ) => {

      // TODO check folder integrity?

      const { integrity } = await readJSON(
        path.join( folder, "node_modules", name, ".qpm" )
      );

      if ( integrity === undefined ) {
        throw error( `It seems that ${pkg} is not installed.` );
      }

      if ( integrity !== hash ) {
        throw error(
          `Expected integrity ${toStr( hash )} but found ${integrity} for ${pkg}.`
        );
      }

    } )
  );
}

export class Checker {

  +reporter: CheckReporter;

  constructor( reporter: CheckReporter ) {
    this.reporter = reporter;
  }

  async check( folder: string ) {

    this.reporter.start();

    const [ pkg, lockfile ] = await Promise.all( [ readPkg( folder ), readLockfile( folder ) ] );

    if ( !shouldReuseLockfile( lockfile ) ) {
      this.reporter.warning( {
        code: "LOCKFILE_NOT_FOUND",
        message: "Lockfile not found."
      } );
      return;
    }

    const dependencies = pkg.dependencies || {};
    const devDependencies = pkg.devDependencies || {};

    this.reporter.comparing( "dependencies" );

    compare( dependencies, lockfile.deps, "dependencies" );

    this.reporter.comparing( "devDependencies" );

    compare( devDependencies, lockfile.devDeps, "devDependencies" );

    this.reporter.integrity();

    await integrity( folder, lockfile );
  }

}

export default function( folder: string ) {
  const reporter = new CheckReporter();
  const checker = new Checker( reporter );
  return checker.check( folder ).then( () => {
    reporter.done();
  }, err => {
    reporter.error( err );
  } );
}
