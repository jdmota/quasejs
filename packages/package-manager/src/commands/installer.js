// @flow
import reporter from "../reporters/installer";
import type { Name, Resolved, Options, Warning } from "../types";
import { mapGet } from "../utils";
import { read as readPkg, validate as validatePkg } from "../pkg";
import {
  type Lockfile,
  shouldReuse as shouldReuseLockfile,
  create as createLockfile,
  read as readLockfile,
  write as writeLockfile
} from "../lockfile";
import { ResolutionSet, Tree } from "../resolution";
import Store from "../store";
import { Resolver } from "../resolve";

const { EventEmitter } = require( "events" );

export class Installer extends EventEmitter {

  +tree: Tree;
  +rootDeps: ResolutionSet;
  +store: Store;
  +opts: Options;
  +warn: Warning => void;
  reuseLockfile: boolean;

  constructor( opts: Options ) {
    super();
    this.opts = opts;
    this.warn = ( warning: Warning ) => {
      this.emit( "warning", warning );
    };
    this.store = new Store( opts, this.warn );
    this.tree = new Tree();
    this.rootDeps = new ResolutionSet();
    this.reuseLockfile = false;
  }

  async resolution( pkg: Object, lockfile: Lockfile, newLockfile: Lockfile ) {
    this.emit( "resolutionStart" );
    const resolver = new Resolver( this, pkg, lockfile, newLockfile );
    await resolver.start();
  }

  async extraction() {

    const extractions = [];
    this.tree.forEach( resolution => {
      extractions.push( this.store.extract( resolution ).then( () => this.emit( "extractionUpdate" ) ) );
    } );

    this.emit( "extractionStart", extractions.length );

    for ( const p of extractions ) {
      await p;
    }
  }

  async linking() {

    const links = [];
    this.tree.forEach( resolution => {
      links.push( this.store.linkResolutionDeps( resolution ).then( () => this.emit( "linkingUpdate" ) ) );
    } );

    this.emit( "linkingStart", links.length );

    for ( const p of links ) {
      await p;
    }
  }

  async localLinking() {
    this.emit( "localLinkingStart", 1 );
    await this.store.linkNodeModules( this.opts.folder, this.rootDeps );
    this.emit( "localLinkingUpdate" );
  }

  async updateLockfile( newLockfile: Lockfile ) {

    this.emit( "updateLockfile" );

    const map: Map<Resolved, number> = new Map();
    this.tree.generate( newLockfile.resolutions, map );

    for ( const nameStr in newLockfile.deps ) {
      // $FlowIgnore
      const name: Name = nameStr;

      newLockfile.deps[ name ].i = mapGet( map, newLockfile.deps[ name ].resolved );
    }

    for ( const nameStr in newLockfile.devDeps ) {
      // $FlowIgnore
      const name: Name = nameStr;

      newLockfile.devDeps[ name ].i = mapGet( map, newLockfile.devDeps[ name ].resolved );
    }

    await writeLockfile( this.opts.folder, newLockfile );
  }

  async install() {

    const { opts } = this;
    const [ pkg, lockfile ] = await Promise.all( [ readPkg( opts.folder ), readLockfile( opts.folder ) ] );

    validatePkg( pkg );

    const reuseLockfile = this.reuseLockfile = !opts.update && shouldReuseLockfile( lockfile );

    this.emit( "folder", { folder: opts.folder } );
    this.emit( "lockfile", { reusing: reuseLockfile } );
    this.emit( "start" );

    const newLockfile = createLockfile();

    await this.resolution( pkg, lockfile, newLockfile );
    await this.extraction();
    await this.linking();
    await this.localLinking();
    await this.updateLockfile( newLockfile );
  }

}

export default function( opts: Options ) {
  const installer = new Installer( opts );
  reporter( installer );
  return installer.install().then( () => {
    installer.emit( "done" );
  }, err => {
    installer.emit( "error", err );
  } );
}
