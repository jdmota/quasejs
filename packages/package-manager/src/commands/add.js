// @flow
import type { Options } from "../types";
import { error } from "../utils";
import { read, readGlobal, write, add, normalizeType } from "../pkg";
import installer from "./installer";

// $FlowIgnore
const latestVersion = require( "latest-version" );
const semver = require( "semver" );

async function inputMapper( required: string ): Promise<{ name: string, version: string }> {
  const i = required.lastIndexOf( "@" );

  if ( i <= 0 ) {
    const name = required.trim();

    return {
      name,
      version: await latestVersion( name )
    };
  }

  const name = required.slice( 0, i ).trim();
  const version = required.slice( i + 1 ).trim();

  const validatedRange = semver.validRange( version );

  if ( validatedRange == null ) {
    throw error( `Invalid version/range for ${name}` );
  }

  return {
    name,
    version: validatedRange
  };
}

/* eslint-disable no-console */

export default async function( options: Options, input: string[] ) {
  const pkg = options.global ? await readGlobal( options.folder ) : await read( options.folder );
  const packages = await Promise.all( input.map( inputMapper ) );
  const added = add( pkg, packages, normalizeType( options.type ) );
  if ( added.length > 0 ) {
    await write( options.folder, pkg );

    console.log( "Adding:" );
    for ( const { name, version } of added ) {
      console.log( `  ${name}@${version}` );
    }
    console.log( "" );
  } else {
    throw error( "Nothing was added" );
  }
  return installer( options );
}
