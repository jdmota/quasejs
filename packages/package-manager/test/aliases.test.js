import { Installer } from "../src/commands/installer";
import { InstallReporterNoop } from "../src/reporters/installer";
import { testProcess, store } from "./utils";

const fs = require( "fs-extra" );
const path = require( "path" );

// $FlowIgnore
jest.setTimeout( 30000 );

describe( "aliases", () => {

  it( "basic", async() => {

    async function install( moreOpts = {} ) {
      await fs.remove( path.resolve( __dirname, "test-folders/aliases/node_modules" ) );
      await new Installer( {
        preferOffline: true,
        folder: path.resolve( __dirname, "test-folders/aliases" ),
        cache: path.join( store, "cache" ),
        store,
        ...moreOpts
      }, new InstallReporterNoop() ).install();

      const processOutput = await testProcess( path.resolve( __dirname, "test-folders/aliases/index.js" ) );
      const lockfile = await fs.readFile( path.resolve( __dirname, "test-folders/aliases/qpm-lockfile.json" ), "utf8" );
      const bins = await fs.readdir( path.resolve( __dirname, "test-folders/aliases/node_modules/.bin" ) );

      expect( processOutput ).toMatchSnapshot();
      expect( lockfile ).toMatchSnapshot();
      expect( bins ).toMatchSnapshot();
    }

    await fs.remove( path.resolve( __dirname, "test-folders/aliases/qpm-lockfile.json" ) );
    await install();
    await fs.remove( path.resolve( __dirname, "test-folders/aliases/qpm-lockfile.json" ) );

  } );

} );
