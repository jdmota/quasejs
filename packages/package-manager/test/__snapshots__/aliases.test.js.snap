// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`installer basic 1`] = `
"{ [Function] sync: [Function] }
{ [Function] sync: [Function] }
{ name: 'read-pkg',
  version: '2.0.0',
  description: 'Read a package.json file',
  license: 'MIT',
  repository: 'sindresorhus/read-pkg',
  author:
   { name: 'Sindre Sorhus',
     email: 'sindresorhus@gmail.com',
     url: 'sindresorhus.com' },
  engines: { node: '>=4' },
  scripts: { test: 'xo && ava' },
  files: [ 'index.js' ],
  keywords:
   [ 'json',
     'read',
     'parse',
     'file',
     'fs',
     'graceful',
     'load',
     'pkg',
     'package',
     'normalize' ],
  dependencies:
   { 'load-json-file': '^2.0.0',
     'normalize-package-data': '^2.3.2',
     'path-type': '^2.0.0' },
  devDependencies: { ava: '*', xo: '*' },
  xo: { esnext: true },
  _resolved: '',
  _integrity: 'sha1-jvHAYjxqbbDcZxPEv6xGMysjaPg=',
  _from: 'https://registry.npmjs.org/read-pkg/-/read-pkg-2.0.0.tgz' }
{ name: 'write-pkg',
  version: '3.1.0',
  description: 'Write a package.json file',
  license: 'MIT',
  repository: 'sindresorhus/write-pkg',
  author:
   { name: 'Sindre Sorhus',
     email: 'sindresorhus@gmail.com',
     url: 'sindresorhus.com' },
  engines: { node: '>=4' },
  scripts: { test: 'xo && ava' },
  files: [ 'index.js' ],
  keywords:
   [ 'json',
     'write',
     'stringify',
     'file',
     'fs',
     'graceful',
     'pkg',
     'package' ],
  dependencies: { 'sort-keys': '^2.0.0', 'write-json-file': '^2.2.0' },
  devDependencies:
   { ava: '*', 'read-pkg': '^2.0.0', tempfile: '^2.0.0', xo: '*' },
  _resolved: '',
  _integrity: 'sha1-AwqZlMyZk9JbTnWp8aGSNgcpHOk=',
  _from: 'https://registry.npmjs.org/write-pkg/-/write-pkg-3.1.0.tgz' }

child process exited with code 0"
`;

exports[`installer basic 2`] = `
"{
	\\"v\\": \\"1\\",
	\\"resolutions\\": [
		[
			\\"builtin-modules\\",
			\\"1.1.1\\",
			\\"https://registry.npmjs.org/builtin-modules/-/builtin-modules-1.1.1.tgz\\",
			\\"sha1-Jw8HbFpywC9bZaR9+Uxf46J4iS8=\\",
			{}
		],
		[
			\\"detect-indent\\",
			\\"5.0.0\\",
			\\"https://registry.npmjs.org/detect-indent/-/detect-indent-5.0.0.tgz\\",
			\\"sha1-OHHMCmoALow+Wzz38zYmRnXwa50=\\",
			{}
		],
		[
			\\"error-ex\\",
			\\"1.3.2\\",
			\\"https://registry.npmjs.org/error-ex/-/error-ex-1.3.2.tgz\\",
			\\"sha512-7dFHNmqeFSEt2ZBsCriorKnn3Z2pj+fd9kmI6QoWw4//DL+icEBfc0U7qJCisqrTsKTjw4fNFy2pW9OqStD84g==\\",
			{
				\\"is-arrayish\\": 3
			}
		],
		[
			\\"is-arrayish\\",
			\\"0.2.1\\",
			\\"https://registry.npmjs.org/is-arrayish/-/is-arrayish-0.2.1.tgz\\",
			\\"sha1-d8mYQFJ6qOyxqLppe4BkWnqSap0=\\",
			{}
		],
		[
			\\"graceful-fs\\",
			\\"4.1.11\\",
			\\"https://registry.npmjs.org/graceful-fs/-/graceful-fs-4.1.11.tgz\\",
			\\"sha1-Dovf5NHduIVNZOBOp8AOKgJuVlg=\\",
			{}
		],
		[
			\\"hosted-git-info\\",
			\\"2.7.1\\",
			\\"https://registry.npmjs.org/hosted-git-info/-/hosted-git-info-2.7.1.tgz\\",
			\\"sha512-7T/BxH19zbcCTa8XkMlbK5lTo1WtgkFi3GvdWEyNuc4Vex7/9Dqbnpsf4JMydcfj9HCg4zUWFTL3Za6lapg5/w==\\",
			{}
		],
		[
			\\"imurmurhash\\",
			\\"0.1.4\\",
			\\"https://registry.npmjs.org/imurmurhash/-/imurmurhash-0.1.4.tgz\\",
			\\"sha1-khi5srkoojixPcT7a21XbyMUU+o=\\",
			{}
		],
		[
			\\"is-builtin-module\\",
			\\"1.0.0\\",
			\\"https://registry.npmjs.org/is-builtin-module/-/is-builtin-module-1.0.0.tgz\\",
			\\"sha1-VAVy0096wxGfj3bDDLwbHgN6/74=\\",
			{
				\\"builtin-modules\\": 0
			}
		],
		[
			\\"is-plain-obj\\",
			\\"1.1.0\\",
			\\"https://registry.npmjs.org/is-plain-obj/-/is-plain-obj-1.1.0.tgz\\",
			\\"sha1-caUMhCnfync8kqOQpKA7OfzVHT4=\\",
			{}
		],
		[
			\\"load-json-file\\",
			\\"2.0.0\\",
			\\"https://registry.npmjs.org/load-json-file/-/load-json-file-2.0.0.tgz\\",
			\\"sha1-eUfkIUmvgNaWy/eXvKq8/h/inKg=\\",
			{
				\\"graceful-fs\\": 4,
				\\"parse-json\\": 10,
				\\"pify\\": 11,
				\\"strip-bom\\": 12
			}
		],
		[
			\\"parse-json\\",
			\\"2.2.0\\",
			\\"https://registry.npmjs.org/parse-json/-/parse-json-2.2.0.tgz\\",
			\\"sha1-9ID0BDTvgHQfhGkJn43qGPVaTck=\\",
			{
				\\"error-ex\\": 2
			}
		],
		[
			\\"pify\\",
			\\"2.3.0\\",
			\\"https://registry.npmjs.org/pify/-/pify-2.3.0.tgz\\",
			\\"sha1-7RQaasBDqEnqWISY59yosVMw6Qw=\\",
			{}
		],
		[
			\\"strip-bom\\",
			\\"3.0.0\\",
			\\"https://registry.npmjs.org/strip-bom/-/strip-bom-3.0.0.tgz\\",
			\\"sha1-IzTBjpx1n3vdVv3vfprj1YjmjtM=\\",
			{}
		],
		[
			\\"lock-cli\\",
			\\"1.2.0\\",
			\\"https://registry.npmjs.org/lock-cli/-/lock-cli-1.2.0.tgz\\",
			\\"sha1-oFCUUw4cXLH+Zy44mba+PgoZUNI=\\",
			{
				\\"lock-system\\": 14
			}
		],
		[
			\\"lock-system\\",
			\\"1.3.0\\",
			\\"https://registry.npmjs.org/lock-system/-/lock-system-1.3.0.tgz\\",
			\\"sha512-gSMfmR+PygNDe5FFN9FNguRT1oyqE05XVCL0Pn62tNqe8X2CfDDBNkEF67lqOlWQ3C2u6hJ4YxAq8cyJ0EO9KQ==\\",
			{}
		],
		[
			\\"make-dir\\",
			\\"1.3.0\\",
			\\"https://registry.npmjs.org/make-dir/-/make-dir-1.3.0.tgz\\",
			\\"sha512-2w31R7SJtieJJnQtGc7RVL2StM2vGYVfqUOvUDxH6bC6aJTxPxTF0GnIgCyu7tjockiUWAYQRbxa7vKn34s5sQ==\\",
			{
				\\"pify\\": 16
			}
		],
		[
			\\"pify\\",
			\\"3.0.0\\",
			\\"https://registry.npmjs.org/pify/-/pify-3.0.0.tgz\\",
			\\"sha1-5aSs0sEB/fPZpNB/DbxNtJ3SgXY=\\",
			{}
		],
		[
			\\"normalize-package-data\\",
			\\"2.4.0\\",
			\\"https://registry.npmjs.org/normalize-package-data/-/normalize-package-data-2.4.0.tgz\\",
			\\"sha512-9jjUFbTPfEy3R/ad/2oNbKtW9Hgovl5O1FvFWKkKblNXoN/Oou6+9+KKohPK13Yc3/TyunyWhJp6gvRNR/PPAw==\\",
			{
				\\"hosted-git-info\\": 5,
				\\"is-builtin-module\\": 7,
				\\"semver\\": 18,
				\\"validate-npm-package-license\\": 19
			}
		],
		[
			\\"semver\\",
			\\"5.5.1\\",
			\\"https://registry.npmjs.org/semver/-/semver-5.5.1.tgz\\",
			\\"sha512-PqpAxfrEhlSUWge8dwIp4tZnQ25DIOthpiaHNIthsjEFQD6EvqUKUDM7L8O2rShkFccYo1VjJR0coWfNkCubRw==\\",
			{}
		],
		[
			\\"validate-npm-package-license\\",
			\\"3.0.4\\",
			\\"https://registry.npmjs.org/validate-npm-package-license/-/validate-npm-package-license-3.0.4.tgz\\",
			\\"sha512-DpKm2Ui/xN7/HQKCtpZxoRWBhZ9Z0kqtygG8XCgNQ8ZlDnxuQmWhj566j8fN4Cu3/JmbhsDo7fcAJq4s9h27Ew==\\",
			{
				\\"spdx-correct\\": 20,
				\\"spdx-expression-parse\\": 21
			}
		],
		[
			\\"spdx-correct\\",
			\\"3.0.0\\",
			\\"https://registry.npmjs.org/spdx-correct/-/spdx-correct-3.0.0.tgz\\",
			\\"sha512-N19o9z5cEyc8yQQPukRCZ9EUmb4HUpnrmaL/fxS2pBo2jbfcFRVuFZ/oFC+vZz0MNNk0h80iMn5/S6qGZOL5+g==\\",
			{
				\\"spdx-expression-parse\\": 21,
				\\"spdx-license-ids\\": 23
			}
		],
		[
			\\"spdx-expression-parse\\",
			\\"3.0.0\\",
			\\"https://registry.npmjs.org/spdx-expression-parse/-/spdx-expression-parse-3.0.0.tgz\\",
			\\"sha512-Yg6D3XpRD4kkOmTpdgbUiEJFKghJH03fiC1OPll5h/0sO6neh2jqRDVHOQ4o/LMea0tgCkbMgea5ip/e+MkWyg==\\",
			{
				\\"spdx-exceptions\\": 22,
				\\"spdx-license-ids\\": 23
			}
		],
		[
			\\"spdx-exceptions\\",
			\\"2.1.0\\",
			\\"https://registry.npmjs.org/spdx-exceptions/-/spdx-exceptions-2.1.0.tgz\\",
			\\"sha512-4K1NsmrlCU1JJgUrtgEeTVyfx8VaYea9J9LvARxhbHtVtohPs/gFGG5yy49beySjlIMhhXZ4QqujIZEfS4l6Cg==\\",
			{}
		],
		[
			\\"spdx-license-ids\\",
			\\"3.0.0\\",
			\\"https://registry.npmjs.org/spdx-license-ids/-/spdx-license-ids-3.0.0.tgz\\",
			\\"sha512-2+EPwgbnmOIl8HjGBXXMd9NAu02vLjOO1nWw4kmeRDFyHn+M/ETfHxQUK0oXg8ctgVnl9t3rosNVsZ1jG61nDA==\\",
			{}
		],
		[
			\\"path-type\\",
			\\"2.0.0\\",
			\\"https://registry.npmjs.org/path-type/-/path-type-2.0.0.tgz\\",
			\\"sha1-8BLMuEFbcJb8LaoQVMPXI4lZTHM=\\",
			{
				\\"pify\\": 11
			}
		],
		[
			\\"read-pkg\\",
			\\"2.0.0\\",
			\\"https://registry.npmjs.org/read-pkg/-/read-pkg-2.0.0.tgz\\",
			\\"sha1-jvHAYjxqbbDcZxPEv6xGMysjaPg=\\",
			{
				\\"load-json-file\\": 9,
				\\"normalize-package-data\\": 17,
				\\"path-type\\": 24
			}
		],
		[
			\\"signal-exit\\",
			\\"3.0.2\\",
			\\"https://registry.npmjs.org/signal-exit/-/signal-exit-3.0.2.tgz\\",
			\\"sha1-tf3AjxKH6hF4Yo5BXiUTK3NkbG0=\\",
			{}
		],
		[
			\\"sort-keys\\",
			\\"2.0.0\\",
			\\"https://registry.npmjs.org/sort-keys/-/sort-keys-2.0.0.tgz\\",
			\\"sha1-ZYU1WEhh7JfXMNbPQYIuH1ZoQSg=\\",
			{
				\\"is-plain-obj\\": 8
			}
		],
		[
			\\"write-file-atomic\\",
			\\"2.3.0\\",
			\\"https://registry.npmjs.org/write-file-atomic/-/write-file-atomic-2.3.0.tgz\\",
			\\"sha512-xuPeK4OdjWqtfi59ylvVL0Yn35SF3zgcAcv7rBPFHVaEapaDr4GdGgm3j7ckTwH9wHL7fGmgfAnb0+THrHb8tA==\\",
			{
				\\"graceful-fs\\": 4,
				\\"imurmurhash\\": 6,
				\\"signal-exit\\": 26
			}
		],
		[
			\\"write-json-file\\",
			\\"2.3.0\\",
			\\"https://registry.npmjs.org/write-json-file/-/write-json-file-2.3.0.tgz\\",
			\\"sha1-K2TIozAE1UuGmMdtWFp3zrYdoy8=\\",
			{
				\\"detect-indent\\": 1,
				\\"graceful-fs\\": 4,
				\\"make-dir\\": 15,
				\\"pify\\": 16,
				\\"sort-keys\\": 27,
				\\"write-file-atomic\\": 28
			}
		],
		[
			\\"write-pkg\\",
			\\"3.1.0\\",
			\\"https://registry.npmjs.org/write-pkg/-/write-pkg-3.1.0.tgz\\",
			\\"sha1-AwqZlMyZk9JbTnWp8aGSNgcpHOk=\\",
			{
				\\"sort-keys\\": 27,
				\\"write-json-file\\": 29
			}
		]
	],
	\\"deps\\": {
		\\"read\\": {
			\\"spec\\": \\"npm:read-pkg@2.0.0\\",
			\\"resolved\\": \\"https://registry.npmjs.org/read-pkg/-/read-pkg-2.0.0.tgz\\",
			\\"i\\": 25
		}
	},
	\\"devDeps\\": {
		\\"lock\\": {
			\\"spec\\": \\"npm:lock-cli@1.2.0\\",
			\\"resolved\\": \\"https://registry.npmjs.org/lock-cli/-/lock-cli-1.2.0.tgz\\",
			\\"i\\": 13
		},
		\\"write\\": {
			\\"spec\\": \\"npm:write-pkg@3.1.0\\",
			\\"resolved\\": \\"https://registry.npmjs.org/write-pkg/-/write-pkg-3.1.0.tgz\\",
			\\"i\\": 30
		}
	},
	\\"optionalDeps\\": {}
}
"
`;

exports[`installer basic 3`] = `
Array [
  "lock",
  "lock.cmd",
  "lock.ps1",
]
`;
